neps/nep-0483.md

```diff
@@ -0,0 +1,83 @@
+
+## Rationale and alternatives
+
+In NEAR, users sign transactions that are later relayed to the network. At the moment of being relayed, the transactions are simple streams of bytes that a node then parses into a [valid transaction](https://nomicon.io/RuntimeSpec/Transactions).
+
+As the protocol expands, more messages will be added for users to sign. We want users to be able to use the protocol safely, making sure that all those signed messages cannot be interpreted as a valid transaction.
+
+In practice, what users sign is the `sha256` hash of a [borsh-serialized](https://borsh.io) structure. This is:
+
+```ts
+sha256.hash(Borsh.serialize(MessageStructure))
+```
+
+Including a large `u32` prefix allows to easily distinguish transactions from messages, ensuring that they never overlap.
+
+```ts
+sha256.hash(Borsh.serialize<u32>(large-number) + Borsh.serialize(MessageStructure))
+```
+
+This is because the first field of a [valid transaction](https://nomicon.io/RuntimeSpec/Transactions) is an `AccountId`, which is encoded as `string length <u32>` + `string chars <u32[]>`. Prepending a prefix in the range of $(2^{30}, 2^{32})$ will automatically make the message deserialize into an invalid transaction, since no `AccountId` has more than 64 chars.
+
+
+## Specification
+
+### On-Chain Messages
+
+All actionable messages, i.e. those that ARE validated on-chain, should prepend an explicit `prefix: u32` with fixed value $2^{30}$ + NEP-number.
+
+For example, if we implement a new actionable message `ActionableMessage` destined to be used on-chain, and the proposal has the number `NEP400`, then the message  must have the following form:
+
+```ts
+// 2**30 + 400 = 1073742224
+sha256.hash(Borsh.serialize<u32>(1073742224) + Borsh.serialize(ActionableMessage))
+```
+
+### Off-Chain Messages
+
+All non-actionable messages, i.e. those that are NOT validated on-chain, should prepend an explicit `prefix: u32` with fixed value $2^{31}$ + NEP-number.
+
+For example, if we implement a new non-actionable message `NonActionableMessage`, destined to be used off-chain, and the proposal has the number `NEP400`, then the message to be signed is:
+
+```ts
+// 2**31 + 400 = 2147484048
+sha256.hash(Borsh.serialize<u32>(2147484048) + Borsh.serialize(NonActionableMessage))
+```
+
+## Security Implications
+
+By implementing this NEP, we can be sure that all new messages do not collide with classic transactions.
+
+## Drawbacks
+
+We will need to update [NEP366](https://github.com/near/NEPs/blob/master/neps/nep-0366.md) to follow this standard, and make sure all future NEPs follow it, which might be hard to track.
+
+If at any point there is an update on how transactions are encoded, this NEP could stop being valid.
```

Comment:
`original comment by` @mfornet 

If any protocol change updates the tx encoding, we should enforce that the first 4 bytes are used following this NEP. (i.e., they must be decoded as a number in the range [0, 2**30).

What do you think the best way is to be alert of this? (Putting some comments in the client source code?)

Reply:
Member
`original comment by` @mfornet

I propose introducing a "protocol update" that enforces the first four bytes of every "classic tx" to be in the proper range. This is the case today (implicitly), but we should add it to the spec, and an extra assertion to the client (no real protocol change required) so we don't break this in the future.